client/tests/integration/smartcontracts/validator_with_custom_token/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+        visit_unregister_peer(Unregister<Peer>),
+
+        // Domain validation
+        visit_set_domain_key_value(SetKeyValue<Domain>),
+        visit_remove_domain_key_value(RemoveKeyValue<Domain>),
+
+        // Account validation
+        visit_unregister_account(Unregister<Account>),
+        visit_mint_account_public_key(Mint<Account, PublicKey>),
+        visit_burn_account_public_key(Burn<Account, PublicKey>),
+        visit_mint_account_signature_check_condition(Mint<Account, SignatureCheckCondition>),
+        visit_set_account_key_value(SetKeyValue<Account>),
+        visit_remove_account_key_value(RemoveKeyValue<Account>),
+
+        // Asset validation
+        visit_register_asset(Register<Asset>),
+        visit_unregister_asset(Unregister<Asset>),
+        visit_mint_asset(Mint<Asset, NumericValue>),
+        visit_burn_asset(Burn<Asset, NumericValue>),
+        visit_transfer_asset(Transfer<Asset, NumericValue, Account>),
+        visit_set_asset_key_value(SetKeyValue<Asset>),
+        visit_remove_asset_key_value(RemoveKeyValue<Asset>),
+
+        // AssetDefinition validation
+        visit_unregister_asset_definition(Unregister<AssetDefinition>),
+        visit_transfer_asset_definition(Transfer<Account, AssetDefinition, Account>),
+        visit_set_asset_definition_key_value(SetKeyValue<AssetDefinition>),
+        visit_remove_asset_definition_key_value(RemoveKeyValue<AssetDefinition>),
+
+        // Permission validation
+        visit_register_permission_token(Register<PermissionTokenDefinition>),
+        visit_grant_account_permission(Grant<Account, PermissionToken>),
+        visit_revoke_account_permission(Revoke<Account, PermissionToken>),
+
+        // Role validation
+        visit_unregister_role(Unregister<Role>),
+        visit_grant_account_role(Grant<Account, RoleId>),
+        visit_revoke_account_role(Revoke<Account, RoleId>),
+
+        // Trigger validation
+        visit_unregister_trigger(Unregister<Trigger<FilterBox, Executable>>),
+        visit_mint_trigger_repetitions(Mint<Trigger<FilterBox, Executable>, u32>),
+        visit_execute_trigger(ExecuteTrigger),
+
+        // Parameter validation
+        visit_set_parameter(SetParameter),
+        visit_new_parameter(NewParameter),
+
+        // Upgrade validation
+        visit_upgrade_validator(Upgrade<iroha_validator::data_model::validator::Validator>),
+    }
+}
+
+impl Validate for CustomValidator {
+    fn verdict(&self) -> &Result {
+        self.0.verdict()
+    }
+    fn deny(&mut self, reason: ValidationFail) {
+        self.0.deny(reason);
+    }
+}
+
+impl ExpressionEvaluator for CustomValidator {
+    fn evaluate<E: Evaluate>(
+        &self,
+        expression: &E,
+    ) -> core::result::Result<E::Value, EvaluationError> {
+        self.0.evaluate(expression)
+    }
+}
+
+/// Entrypoint to return permission token definitions defined in this validator.
+#[entrypoint]
+pub fn permission_tokens() -> Vec<PermissionTokenDefinition> {
```

Comment:
> 2. This entrypoint function should be automatically generated with #[entrypoint] on validate function

I think it would be very unclear from user point of view. Also it would require to know the type of validator, so I'll need to add another attribute thing in this macro

Reply:
> 1. I think this should be a method on the Validate trait

Before we didn't have something like `Validator::validate()` as an entrypoint. It's a custom functions which permforms matching over `NeedsValidationBox`. Maybe we should create this method inside `Validate` trait as well?

Reply:
I mean, I did it in the same way it was before. And we should either do it like that or replace with trait methods, after which entrypoint functions will probably be just wrappers around methods

Reply:
> I think it would be very unclear from user point of view. Also it would require to know the type of validator, so I'll need to add another attribute thing in this macro

What do you mean by type of validator? What types of validators are there?
My point here was that user should not be bothered with writing this export. It should be autogenerated

Reply:
> I don't want to see more bloat in the validator_with_admin

I think the problem is that it was already bloated and we should try to make it more clear in another way. This new entrypoint doesn't add much lines but keep things explicit for user (what I thought we consider to be good).

> What do you mean by type of validator?

I mean that macro doesn't know on which type to call `<_ as Validator>::permission_tokens()`. We should pass this information to the macro

Reply:
> I think the problem is that it was already bloated and we should try to make it more clear in another way. This new entrypoint doesn't add much lines but keep things explicit for user (what I thought we consider to be good).

there is no reason for this to be explicit. If explicit was always preferred we wouldn't have `entrypoint` macro. The problem is that it's hard to hide it

> I mean that macro doesn't know on which type to call <_ as Validator>::permission_tokens(). We should pass this information to the macro

I see, this is a pervasive problem. `ExpressionEvaluator` has the same problem. Perhaps we'll come up with a way to hide these away. Let me close this and make another suggestion